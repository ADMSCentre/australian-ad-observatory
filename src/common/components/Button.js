import React from 'react';
import PropTypes from 'prop-types';
import BButton from 'react-bootstrap/Button';

const openPage = to => {
  const base = 'webpage/index.html';
  const url = to ? `${base}#${to}` : base;
  const extensionUrl = chrome.runtime.getURL(url);

  /* istanbul ignore next */
  return () => {
    window.open(extensionUrl, '_blank');
  };
};

const Button = ({ type, label, to, variant, onClick }) => {
  const doClick = onClick || openPage(to);

  return type === 'link' ? (
    <a href="#" onClick={doClick}>
      {label}
    </a>
  ) : (
    <BButton variant={variant} block onClick={doClick}>
      {label}
    </BButton>
  );
};
Button.displayName = 'Button';

Button.propTypes = {
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
  type: PropTypes.oneOf(['link', 'button']),
  to: PropTypes.string, // you must have either a 'to' or 'onClick'
  onClick: PropTypes.func,
  variant: PropTypes.oneOf([
    'primary',
    'secondary',
    'success',
    'warning',
    'danger',
    'info',
    'light',
    'dark',
    'link'
  ])
};

Button.defaultProps = {
  type: 'button',
  to: null,
  onClick: null,
  variant: undefined
};

export default Button;
