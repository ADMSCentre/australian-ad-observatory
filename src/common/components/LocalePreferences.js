import React, { Fragment } from 'react';
import PreferenceSelector from './PreferenceSelector';
import { PREF_COUNTRY, PREF_LANGUAGE, PREF_AGE, PREF_GENDER } from 'common/keys';
import { COUNTRIES } from 'common/countries';
import { LANGUAGES } from 'common/languages';
import { AGES } from 'common/ages';
import { GENDERS } from 'common/genders';

const toArray = object => Object.keys(object).map(value => ({ value, text: object[value] }));

const LocalePreferences = () => (
  <Fragment>
    <PreferenceSelector
      options={toArray(COUNTRIES)}
      label="Your Country"
      storageKey={PREF_COUNTRY}
    />
    <PreferenceSelector
      options={toArray(LANGUAGES)}
      label="Your Language"
      storageKey={PREF_LANGUAGE}
    />
    <PreferenceSelector options={toArray(AGES)} label="Age" storageKey={PREF_AGE} />
    <PreferenceSelector options={toArray(GENDERS)} label="Gender" storageKey={PREF_GENDER} />
  </Fragment>
);
LocalePreferences.displayName = 'Locale preferences';

export default LocalePreferences;
