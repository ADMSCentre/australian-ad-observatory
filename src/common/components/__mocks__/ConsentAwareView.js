/* eslint-disable react/prop-types */
import React, { Fragment } from 'react';
import proceed from 'common/utils/proceed';

let granted;

export const setGranted = gtd => {
  granted = gtd;
};

const ConsentAwareView = ({ children, whenGrantedIs, whenGrantedIsNot }) =>
  proceed(granted, whenGrantedIs, whenGrantedIsNot) ? <Fragment>{children}</Fragment> : null;

export default ConsentAwareView;
