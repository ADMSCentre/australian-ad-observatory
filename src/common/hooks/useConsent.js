import { useCallback, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import actions from 'common/state/consent/actions';

const useConsent = () => {
  const { granted, acceptedAt, viewedAt } = useSelector(
    ({ consent: { granted, acceptedAt, viewedAt } }) => ({
      granted,
      acceptedAt,
      viewedAt
    })
  );

  const dispatch = useDispatch();

  const loadConsent = useCallback(() => dispatch(actions.loadConsent()), [dispatch]);

  const saveConsent = useCallback(() => dispatch(actions.saveConsent()), [dispatch]);

  const viewedConsent = useCallback(() => dispatch(actions.viewedConsent()), [dispatch]);

  useEffect(() => {
    /* istanbul ignore else */
    if (!granted) loadConsent();
  }, [granted]);

  return { granted, acceptedAt, viewedAt, saveConsent, viewedConsent };
};

export default useConsent;
