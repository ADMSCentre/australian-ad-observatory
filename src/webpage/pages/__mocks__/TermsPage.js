import React from 'react';

const TermsPage = () => <h1>Terms Page goes here</h1>;

export default TermsPage;
