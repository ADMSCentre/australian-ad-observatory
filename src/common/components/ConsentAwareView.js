import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import useConsent from 'common/hooks/useConsent';
import proceed from 'common/utils/proceed';

const ConsentAwareView = ({ children, whenGrantedIs, whenGrantedIsNot }) => {
  const { granted } = useConsent();

  return proceed(granted, whenGrantedIs, whenGrantedIsNot) ? <Fragment>{children}</Fragment> : null;
};
ConsentAwareView.displayName = 'Consent aware view';

ConsentAwareView.propTypes = {
  children: PropTypes.node.isRequired,
  whenGrantedIs: PropTypes.oneOf(['none', 'old', 'current']),
  whenGrantedIsNot: PropTypes.oneOf(['none', 'old', 'current'])
};

ConsentAwareView.defaultProps = {
  whenGrantedIs: null,
  whenGrantedIsNot: null
};

export default ConsentAwareView;
