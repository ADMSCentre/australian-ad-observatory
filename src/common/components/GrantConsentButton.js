import React from 'react';
import PropTypes from 'prop-types';
import Button from './Button';

import useConsent from 'common/hooks/useConsent';
import useNavigation from 'common/hooks/useNavigation';

const GrantConsentButton = ({ label, to }) => {
  const { granted, saveConsent } = useConsent();
  const { goto } = useNavigation();

  const saveAndGo = () => {
    saveConsent();
    if (to) goto(to);
  };

  return granted === null || granted === 'current' ? null : (
    <Button variant="primary" onClick={saveAndGo} label={label} />
  );
};
GrantConsentButton.displayName = 'Grant aware button';

GrantConsentButton.propTypes = {
  label: PropTypes.string.isRequired,
  to: PropTypes.string
};

GrantConsentButton.defaultProps = {
  to: null
};

export default GrantConsentButton;
