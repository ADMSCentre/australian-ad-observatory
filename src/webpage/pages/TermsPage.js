import React, { Fragment, useEffect } from 'react';
import ConsentAwareView from 'common/components/ConsentAwareView';
import GrantConsentButton from 'common/components/GrantConsentButton';
import useConsent from 'common/hooks/useConsent';

const TermsPage = () => {
  const { viewedConsent } = useConsent();

  useEffect(() => {
    /* istanbul ignore else */
    viewedConsent();
  }, []);

  return (
    <Fragment>
      <header>
        <h1>Terms of Use</h1>
      </header>
      <div dangerouslySetInnerHTML={{ __html: process.env.CONSENT_TEXT }} />
      <ConsentAwareView whenGrantedIs="none">
        <p>
          You need to accept the above terms and conditions before you can use the browser
          extension.
        </p>
        <GrantConsentButton label="I accept these terms and conditions" to="preferences" />
      </ConsentAwareView>
      <ConsentAwareView whenGrantedIs="old">
        <p>The above terms and conditions have been updated since you last agreed to them.</p>
        <GrantConsentButton label="I accept the updated terms and conditions" to="preferences" />
      </ConsentAwareView>
      <ConsentAwareView whenGrantedIs="current">
        <p>You accepted the above terms and conditions.</p>
      </ConsentAwareView>
    </Fragment>
  );
};
TermsPage.displayName = 'TermsPage';

export default TermsPage;
