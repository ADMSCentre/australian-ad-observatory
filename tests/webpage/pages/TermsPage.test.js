import React from 'react';
import renderer from 'react-test-renderer';
import TermsPage from 'webpage/pages/TermsPage';

import useConsent from 'common/hooks/useConsent';

import { setGranted } from 'common/components/ConsentAwareView';

jest.mock('common/components/ConsentAwareView');
jest.mock('common/components/GrantConsentButton');
jest.mock('common/hooks/useConsent');

['none', 'old', 'current'].forEach(granted => {
  describe(`when consent is '${granted}'`, () => {
    let tree;

    const viewedConsent = jest.fn();

    beforeAll(() => {
      useConsent.mockReturnValue({ viewedConsent });
      setGranted(granted);
      tree = renderer.create(<TermsPage />);
    });

    it('rendered correctly', () => {
      expect(tree).toMatchSnapshot();
    });
  });
});
