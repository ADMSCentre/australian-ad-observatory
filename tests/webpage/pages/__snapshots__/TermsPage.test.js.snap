// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when consent is 'current' rendered correctly 1`] = `
Array [
  <header>
    <h1>
      Terms of Use
    </h1>
  </header>,
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<p>I consent</p>",
      }
    }
  />,
  <p>
    You accepted the above terms and conditions.
  </p>,
]
`;

exports[`when consent is 'none' rendered correctly 1`] = `
Array [
  <header>
    <h1>
      Terms of Use
    </h1>
  </header>,
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<p>I consent</p>",
      }
    }
  />,
  <p>
    You need to accept the above terms and conditions before you can use the browser extension.
  </p>,
  <button>
    I accept these terms and conditions
  </button>,
]
`;

exports[`when consent is 'old' rendered correctly 1`] = `
Array [
  <header>
    <h1>
      Terms of Use
    </h1>
  </header>,
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<p>I consent</p>",
      }
    }
  />,
  <p>
    The above terms and conditions have been updated since you last agreed to them.
  </p>,
  <button>
    I accept the updated terms and conditions
  </button>,
]
`;
